src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java

```diff
@@ -377,35 +377,22 @@ private Acl createAcl(DocId docId, String parentFragment,
   private void setMetadata(Document document, Response response) {
     IDocumentProperties properties =
         options.getObjectFactory().getDocumentProperties(document);
-    for (String name : getPropertyNames(properties)) {
-      ArrayList<String> list = new ArrayList<>();
-      properties.getProperty(name, list);
-      for (String value : list) {
-        response.addMetadata(name, value);
-      }
-    }
-  }
-
-  private Set<String> getPropertyNames(IDocumentProperties properties)  {
-    Set<String> names = new HashSet<String>();
-    for (String property : properties.getPropertyNames()) {
-      if (property != null) {
-        if (options.getIncludedMetadata().size() != 0) {
-          // includeMeta - excludeMeta
-          if ((!options.getExcludedMetadata().contains(property)
-              && options.getIncludedMetadata().contains(property))) {
-            names.add(property);
-          }
-        } else {
-          // superSet - excludeMeta
-          if ((!options.getExcludedMetadata().contains(property))) {
-            names.add(property);
-          }
+    Set<String> names = new TreeSet<>();
+    for (String name : properties.getPropertyNames()) {
+      if ((options.getIncludedMetadata().isEmpty()
+              || options.getIncludedMetadata().contains(name))
+          && !options.getExcludedMetadata().contains(name)) {
+        ArrayList<String> list = new ArrayList<>();
+        properties.getProperty(name, list);
+        for (String value : list) {
+          response.addMetadata(name, value);
+        }
+        if (!list.isEmpty()) {
```

Comment:
I would have put the for-each loop inside the if, since you are already testing for !isEmpty()

Or, as you suggest, just add the name to the set inside the loop and bear the overhead of multiple adds for multivalued props.